### ISSUE TITLE
how to use num_workers in ddp training?

### ISSUE THREAD

**Issue by zhangvia** [webdataset/webdataset#303]
i'm using the webdataset in ddp training. everything works fine when i set the num_workers 0. but if  num_workers > 0,the total steps of an epoch was wrong.
```python
dat = Webdataset(url,8000,2,True)
# loader = data.DataLoader(dat(),pin_memory=True,shuffle=False,batch_size=20,num_workers=16)
loader = wds.WebLoader(dat(),pin_memory=True,shuffle=False,batch_size=20,num_workers=2)
# loader = DataLoaderX(dat(),pin_memory=True,shuffle=False,batch_size=20)
cnt = 0
for d in loader:
    cnt += 1
    print(cnt)
    # print(d[0].shape)
print(f"loader len is {cnt}")
```
i try three different dataloaders. the DataLoaderX is using prefetch_generator.
the dat() will return :
```python
return wds.WebDataset(self.url,resampled=True).shuffle(1000).map(preprocess_train).with_epoch(self.len // self.world_size)
```
the world_size now was set to 2.

when num_workers=0, the total steps of an epoch is 8000 / 2/20=200,but if the num_workers = 16, the total steps will be 8000 / 2 / 200 * 16 = 3200
why does it happen?
could anyone give me some advice?

**Comment by tmbdev** [comment]
The long and the short of it is that you probably want to remove the with_epoch from the WebDataset and put it onto WebLoader:

```
data = wds.WebDataset(self.url,resampled=True).shuffle(1000).map(preprocess_train)
loader = wds.WebLoader(dat(),pin_memory=True,shuffle=False,batch_size=20,num_workers=2).with_epoch(...)
```

You might also want to add cross-worker shuffling, something like:

```
loader = wds.WebLoader(dat(),pin_memory=True,shuffle=False,batch_size=20,num_workers=2)
loader = loader.unbatched().shuffle(2000).batched(20).with_epoch(200)```
```
